spec.md

```diff
@@ -2961,6 +2961,104 @@ For this reason, it is an error to use `rwprobe`{.firrtl} on any port on a publi
 
 Probes may not be inputs to modules.
 
+# Contracts
+
+The `contract`{.firrtl} declaration provides a mechanism to verify a predicate or property, and to then use this predicate or property to simplify the circuit for further verification or simulation.
+
+``` firrtl,notest
+FIRRTL version 4.2.0
+circuit Foo:
+  public module Foo:
+    input a: UInt<42>
+    input b: UInt<1337>
+    output x: UInt<42>
+    output y: UInt<1337>
+    ;; snippetbegin
+    contract c, d = a, b:
+      ; contract body
+    connect x, c
+    connect y, d
+    ;; snippetend
+```
+
+A contract conceptually acts like a `node`{.firrtl}, defining one or more names and assigning a value to each.
+It also has a block to hold statements that describe the predicates and properties to be verified.
+Most commonly the `circt_verif_require` and `circt_verif_ensure` intrinsics ([@sec:intrinsics]) are used to describe pre- and post-conditions of the contract.
+These intrinsics can only be used in the body block of a contract.
+The types of the defined names are the types of the expressions given in the definition and must be a ground type or a potentially nested aggregate type of only ground types.
+Probes, properties, and bundles with flips cannot be used in a contract.
```

Comment:
Is it that you can't use probe, properties, and bundles with flips or that you aren't allowed to send data out of a contract body?

Put differently, it seems odd that you can't do this:

```
contract a = b:
  wire a: Probe<UInt<1>>
```

However, this is illegal based on how this is written.

Reply:
The example that comes to my mind is something like:
```
wire b: Probe<UInt<1>>
contract a = b:
```
When you use the contract as a cutpoint, this would essentially become this in MLIR:
```
node a = symbolic_value : Probe<UInt<1>>
```
I'm not sure what a symbolic probe would be. This is almost like a pointer in software that is symbolic, potentially pointing anywhere.

You could argue that the standard should allow for this, and that the compiler should just not introduce symbolic values for types such as probes or properties. That might actually be a decent way of dealing with this.